doc/design/networking/Investigation-on-NSX-integration.md

```diff
@@ -0,0 +1,109 @@
+  * NSX manager create an IP sets  `<vch name>-IPSets-network-a` with the subnet of this network (or Security Tag `<vch name>-SecurityTag-network-a` and add it to containerVMs)
+  * NSX manager create a security group `<vch name>-SecurityGroup-network-a` (with members: IP Sets == `<vch name>-IPSets-network-a` or Security Tag == `<vch name>-SecurityTag-network-a`))
+  * NSX manager apply the policy groups to this security group
+
+### docker run -net=network-a
+  * create a containerVM 
+
+![how security groups (with IP Sets as membership condition) and policy groups can be used to isolate container networks] (pics/security-group-with-ip-sets.png)
+
+![how security groups (with Security Tag as membership condition) and policy groups can be used to isolate container networks] (pics/security-group-with-ip-sets.png)
+
+## NSX Manager API:
+![Code flow to use NSX Manager API] (pics/with-nsx.png)
+
+### API endpoint
+https://{nsxmanager's hostname or IP}/api 
+
+### API calls
+(details in http://pubs.vmware.com/nsx-63/topic/com.vmware.ICbase/PDF/nsx_63_api.pdf , note that many NSX API methods reference vCenter object IDs in URI parameters, query parameters, request bodies, and response bodies, also it takes XML as request body (sad) )
+* docker network create
+  * create IP Sets
+     * POST /2.0/services/ipset/{scopeMoref}
+  * or create Security Tag
+     * POST /2.0/services/securitytags/tag
+  * create security group withe IP Sets or Security Tag as members
+     * POST /2.0/services/securitygroup/bulk/{scopeId} (For the scopeId use globalroot-0 for non-universal security groups and universalroot-0 for universal security groups.)
+  * create security policy and apply it to security group
+     * POST /2.0/services/policy/securitypolicy (specify securityGroupBinding in the xml request body)
+
+* docker network delete 
+  * check if any VMs attached to this security group
+     * GET /2.0/services/securitygroup/{objectId}/translation/virtualmachine 
+  * delete the network (security group, policy group, IP Sets)
+     * DELETE /2.0/services/securitytags/tag/{tagId}
+     * DELETE /2.0/services/securitygroup/{objectId} 
+     * DELETE /2.0/services/ipset/{ipsetId}
+
+* docker network list
+  * list all the security groups
+    * GET /2.0/services/securitygroup/internal/scope/{scopeId}
+  * filter security group names by `<vch name>-SecurityGroup` prefix
+  * or if we keep KV pairs of each created network, then we just need to use that to list all the networks
+
+* docker run <image> --net=<network-name>
+  * create a containerVM
+  * if using security tag as the membership condition, need to assign security tag to the containerVM
+    * POST /2.0/services/securitytags/vm/{vmId}
+
+## State Storage
+  * we can map a docker network name to a security group name as `<vch name>-SecurityGroup-<network name>`
+  * we can store KV pairs for each created network and then everything else will be in the NSX control plane
```

Comment:
sorry I was not clear here. I meant storing some key/value pair like: 
  [key]: docker network name, [value]: information of the related objects created on NSX side. 
can help fetching network info faster. 
but as your previous comment mentioned, pulling from infrastructure would be preferred. Then no states need to be stored outside of NSX.